File: Coding_Questions/Python/UnderscorifySubstring.py
# ---Underscorify Substring---

from typing import List


# -----------------Solution 2--------------------------------------
def underscorify_substring(string: str, substring: str) -> str:
    locations_list = get_locations(string, substring)
    collapsed_locations_list = collapse_locations(locations_list)
    string = underscorify(collapsed_locations_list, string)
    return string


def underscorify(collapsed_locations_list: List[List[int]], string: str) \
        -> str:
    underscores_added = 0
    for locations_sub_list in collapsed_locations_list:
        string = string[:locations_sub_list[0] + underscores_added] + "_" + \
                 string[locations_sub_list[0] +
                        underscores_added:locations_sub_list[1] +
                                          underscores_added] + \
                 "_" + string[locations_sub_list[1] + underscores_added:]
        underscores_added += 2

    return string


def get_locations(input_string: str, input_substring: str) -> List[List[int]]:
    locations_list = list()
    current_idx = 0

    while current_idx < len(input_string):
        spot_idx = input_string.find(input_substring, current_idx)
        if spot_idx != -1:
            locations_list.append([spot_idx, spot_idx + len(input_substring)])
            current_idx = spot_idx + 1
        else:
            break

    return locations_list


def collapse_locations(locations_list: List[List[int]]) -> List[List[int]]:
    collapsed_locations_list = list()
    first_time = True

    for location_sub_list in locations_list:
        if first_time is True:
            collapsed_locations_list.append(location_sub_list)
            first_time = False
        else:
            if collapsed_locations_list[
                len(collapsed_locations_list) - 1
            ][1] >= location_sub_list[0]:
                collapsed_locations_list[
                    len(collapsed_locations_list) - 1
                ][1] = location_sub_list[1]
            else:
                collapsed_locations_list.append(location_sub_list)

    return collapsed_locations_list

# -----------------Solution 2--------------------------------------
# def underscorify_substring(string: str, substring: str) -> str:
#     locations_list = build_locations_2d_list(string, substring)
#     collapsed_locations_list = collapse_locations_list(locations_list)
#     return build_underscore_string(collapsed_locations_list, string)
#
#
# def build_underscore_string(locations_list: List[List[int]], string: str) -> str:
#     underscores = 0
#     for sub_list in locations_list:
#         string = string[:sub_list[0] + underscores] + "_" + \
#                  string[sub_list[0] +
#                         underscores:sub_list[1] + underscores] + "_" + \
#                  string[sub_list[1] + underscores:]
#         underscores += 2
#     return string
#
#
# def collapse_locations_list(locations: List[List[int]]) \
#         -> List[List[int]]:
#     if not len(locations):
#         return locations
#     new_locations = [locations[0]]
#     previous = new_locations[0]
#     for i in range(1, len(locations)):
#         current = locations[i]
#         if current[0] <= previous[1]:
#             previous[1] = current[1]
#         else:
#             new_locations.append(current)
#             previous = current
#     return new_locations
#
#
# def build_locations_2d_list(string: str, substring: str) -> List[List[int]]:
#     locations = list()
#     start_idx = 0
#     while start_idx < len(string):
#         next_idx = string.find(substring, start_idx)
#         if next_idx != -1:
#             locations.append([next_idx, next_idx + len(substring)])
#             start_idx = next_idx + 1
#         else:
#             break
#     return locations


actual = underscorify_substring(
    "abababababababababababababaababaaabbababaa", "a"
)
expected = "_a_b_a_b_a_b_a_b_a_b_a_b_a_b_a_b_a_b_a_b_a_b_a_b_a_b_aa_b_a_b" \
           "_aaa_bb_a_b_a_b_aa_"
print(f"EXPECTED:\t{expected}")
print("------------------------")
print(f"ACTUAL:\t\t{actual}")
actual2 = underscorify_substring(
    "testthis is a testtest to see if testestest it works", "test"
)
expected2 = "_test_this is a _testtest_ to see if _testestest_ it works"
print(f"EXPECTED:\t{expected2}")
print("------------------------")
print(f"ACTUAL:\t\t{actual2}")
actual3 = underscorify_substring(
    "ttttttttttttttbtttttctatawtatttttastvb", "ttt"
)
expected3 = "_tttttttttttttt_b_ttttt_ctatawta_ttttt_astvb"
print(f"EXPECTED:\t{expected3}")
print("------------------------")
print(f"ACTUAL:\t\t{actual3}")
assert actual == expected
assert actual2 == expected2
assert actual3 == expected3

File: Coding_Questions/C++/UnderscorifySubstring.cpp
// ---Underscorify Substring---

#include <vector>


std::vector<std::vector<int>> get_locations(const std::string &str, const std::string &substring) {
    std::vector<std::vector<int>> locations_2d_vector;
    int current_idx = 0;

    while (current_idx < str.size()) {
        int spot_idx = str.find(substring, current_idx);
        if (spot_idx != std::string::npos) {
            locations_2d_vector.push_back({spot_idx, spot_idx + (int)substring.size()});
            current_idx = spot_idx + 1;
        } else {
            break;
        }
    }

    return locations_2d_vector;
}


std::vector<std::vector<int>> collapse_vector(const std::vector<std::vector<int>> &vec) {
    std::vector<std::vector<int>> collapsed_vector;
    bool first_run = true;

    for (const auto &location_vec : vec) {
        if (first_run) {
            collapsed_vector.push_back(location_vec);
            first_run = false;
        } else {
            if (collapsed_vector[collapsed_vector.size() - 1][1] >= location_vec[0]) {
                collapsed_vector[collapsed_vector.size() - 1][1] = location_vec[1];
            } else {
                collapsed_vector.push_back(location_vec);
            }
        }
    }

    return collapsed_vector;
}


std::string underscorify(const std::string &str, const std::vector<std::vector<int>> &locations_vec) {
    std::string return_str = str;
    int underscores_added = 0;

    for (const auto &location_vec : locations_vec) {
        return_str = return_str.substr(0, location_vec[0] + underscores_added) + "_" +
                return_str.substr(location_vec[0] + underscores_added, location_vec[1] - location_vec[0]) +
                "_" + return_str.substr(location_vec[1] + underscores_added);
        underscores_added += 2;
    }

    return return_str;
}


std::string underscorify_substring(const std::string &str, const std::string &substring) {
    std::vector<std::vector<int>> locations_2d_vec = get_locations(str, substring);
    std::vector<std::vector<int>> collapsed_locations_2d_vec = collapse_vector(locations_2d_vec);
    std::string return_str = underscorify(str, collapsed_locations_2d_vec);
    return return_str;
}